webapi/src/main/scala/org/knora/webapi/messages/admin/responder/projectsmessages/ProjectsMessagesADM.scala

```diff
@@ -505,130 +506,61 @@ case class ProjectADM(
-
-    if (parametersCount > 1) throw BadRequestException("Only one option allowed for project identifier.")
-
-    new ProjectIdentifierADM(
-      maybeIri =
-        sf.validateAndEscapeOptionalProjectIri(maybeIri, throw BadRequestException(s"Invalid user project $maybeIri")),
-      maybeShortname = sf.validateAndEscapeOptionalProjectShortname(
-        maybeShortname,
-        throw BadRequestException(s"Invalid user project shortname $maybeShortname")
-      ),
-      maybeShortcode = sf.validateAndEscapeOptionalProjectShortcode(
-        maybeShortcode,
-        throw BadRequestException(s"Invalid user project shortcode $maybeShortcode")
-      )
-    )
+  case class Iri(value: String) extends ProjectIdentifierADM {
+    def apply(value: String): ProjectIri = ProjectIri.make(value).fold(e => throw e.head, v => v)
   }
-}
 
-/**
- * Represents the project's identifier. It can be an IRI, shortcode or shortname.
- *
- * @param maybeIri       the project's IRI.
- * @param maybeShortname the project's shortname.
- * @param maybeShortcode the project's shortcode
- */
-class ProjectIdentifierADM private (
-  maybeIri: Option[IRI] = None,
-  maybeShortname: Option[String] = None,
-  maybeShortcode: Option[String] = None
-) {
-
-  // squash and return value.
-  val value: String = List(
-    maybeIri,
-    maybeShortname,
-    maybeShortcode
-  ).flatten.head
-
-  def hasType: ProjectIdentifierType =
-    if (maybeIri.isDefined) {
-      ProjectIdentifierType.IRI
-    } else if (maybeShortcode.isDefined) {
-      ProjectIdentifierType.SHORTCODE
-    } else {
-      ProjectIdentifierType.SHORTNAME
-    }
+  case class Shortcode(value: String) extends ProjectIdentifierADM {
+    def apply(value: String): ShortCode = ShortCode.make(value).fold(e => throw e.head, v => v)
+  }
 
-  /**
-   * Tries to return the value as an IRI.
-   */
-  def toIri: IRI =
-    maybeIri.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.IRI' type.")
-    )
+  case class Shortname(value: String) extends ProjectIdentifierADM {
+    def apply(value: String): ShortName = ShortName.make(value).fold(e => throw e.head, v => v)
+  }
 
-  /**
-   * Returns an optional value of the identifier.
-   */
-  def toIriOption: Option[IRI] =
-    maybeIri
+  case class Uuid(value: String) extends ProjectIdentifierADM
 
-  /**
-   * Tries to return the value as an SHORTNAME.
-   */
-  def toShortname: String =
-    maybeShortname.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.SHORTNAME' type.")
-    )
+  def asIriOption(id: ProjectIdentifierADM): Option[String] =
+    id match {
+      case Iri(value) => Some(value)
+      case _          => None
+    }
 
-  /**
-   * Returns an optional value of the identifier.
-   */
-  def toShortnameOption: Option[String] =
-    maybeShortname
+  def asShortcodeOption(id: ProjectIdentifierADM): Option[String] =
```

Comment:
Please, no parameters here. You want this method to be part of the object, where you can ask the object to return itself as something else.

It should be more something like this:

```
object ProjectIdentifierADM { self =>
     val asShortcodeOption: Option[String] =
           self match {
                 case Shortcode(value) => Some(value)
                 case _                => None
      }
}
```

In the callsite you would then write something like: `identifier.asShortcodeOption`.